### PR TITLE
Issue #98: add tests for src/tools.cpp

### PR DESCRIPTION
**Pull request by asashnov**
Related to #98 (partially fix)

* Add unittests for src/tools.cpp
* Remove some unused code
* Limit the scope of helper functions in src/tools.cpp

### CODE REVIEW

**Comment by asashnov** [comment]
Seems CodeCov is not completely configured.

Crash of `CI / Linux (native_static)` in `commont-tools-test` may be caused by:
```
/usr/src/gtest/src/gtest.cc:3451: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
```

`CI / Linux (native_dyn)` fails to load libicu dynamic library:
```
/home/runner/zim-tools/build/test/commont-tools-test: error while loading shared libraries: libicui18n.so.58: cannot open shared object file: No such file or directory
```

I think those problems are environmental, not in the code.

File: test/commont-tools-test.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
+#include "gtest/gtest.h"
```
Comment (MiguelRocha) [inline test/commont-tools-test.cpp:1]: Typo in the filename

File: test/commont-tools-test.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  str = computeAbsolutePath("/home/alex/oz/zim-tools/test/data", "minimal-content/hello.html");
+  EXPECT_EQ(str, "/home/alex/oz/zim-tools/test/minimal-content/hello.html");
+}
+
+TEST(CommonTools, computeRelativePath)
+{
+  std::string str;
+
+  // FIXME: infinitive loop and memory bomb
+  // res = computeRelativePath("", "");
```
Comment (MiguelRocha) [inline test/commont-tools-test.cpp:54]: Avoid adding commented code. Since your testing this functions and discovering bugs, it is a good opportunity to fix them. We always want to increase the code quality.

File: test/meson.build [hunk 3]
```diff
@@ -11,6 +12,15 @@ zimwriter_srcs = [  '../src/zimwriterfs/article.cpp',
                     '../src/tools.cpp']
 
 if gtest_dep.found() and not meson.is_cross_build()
+
+    test_tools_exe = executable('commont-tools-test',
```
Comment (MiguelRocha) [inline test/meson.build:16]: Why is the icu-i18n lib necessary for the tools?

File: test/testhelpers.cpp [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#include <unistd.h>
+#include "testhelpers.h"
+#include "../src/tools.h"
+
+
+TestHelpers::TestHelpers()
```
Comment (MiguelRocha) [inline test/testhelpers.cpp:6]: I dont see the need for adding this class. Please use a helper function if nedded.

File: test/meson.build (hunk 3 above)
Comment (asashnov) [reply test/meson.build:16]: removeAccents() uses it:
```c
  static std::unique_ptr<icu::Transliterator> removeAccentsTrans(icu::Transliterator::createInstance(
      "Lower; NFD; [:M:] remove; NFC", UTRANS_FORWARD, status));
  icu::UnicodeString ustring(text.c_str());
  removeAccentsTrans->transliterate(ustring);
```

File: test/testhelpers.cpp (hunk 4 above)
Comment (asashnov) [reply test/testhelpers.cpp]: Indeed. I thought I need more flexibility, but I didn't.

File: test/commont-tools-test.cpp (hunk 2 above)
Comment (asashnov) [reply test/commont-tools-test.cpp]: Code fixed, more tests added.

File: test/commont-tools-test.cpp (hunk 1 above)
Comment (asashnov) [reply test/commont-tools-test.cpp]: Renamed: commont-tools-test.cpp => tools-test.cpp

File: test/testhelpers.cpp (hunk 4 above)
Comment (asashnov) [reply test/testhelpers.cpp]: class and files testhelpers.(cpp|h) removed.

File: test/tools-test.cpp [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+#include "../src/tools.h"
+
+static std::string get_tests_data_dir()
+{
+  return "data";
+}
+
+TEST(CommonTools, getFileSize)
+{
+  std::string fn = get_tests_data_dir() + "/minimal-content/favicon.png";
```
Comment (mgautierfr) [inline test/tools-test.cpp:12]: Wouldn't be better to create a file in tmp and test it is correctly detected as existing instead of storing a file in the repository ?

File: test/tools-test.cpp [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+{
+  unsigned char data[] = { 0xff, 0x00, 0x7a };
+  std::string txt = base64_encode(data, sizeof(data));
+  EXPECT_EQ(txt, "/wB6");
+}
+
+TEST(CommonTools, decodeUrl)
+{
+  const char *src = "https://www.example.com/cgi-bin/search.cgi?q=example%20search";
+  std::string res = decodeUrl(src);
```
Comment (mgautierfr) [inline test/tools-test.cpp:28]: We should test more decoding char than `%20` only.

File: test/tools-test.cpp [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  replaceStringInPlace(str, "", "");
+  EXPECT_EQ(str, "");
+
+  str = "abcd";
+  replaceStringInPlace(str, "a", "b");
+  EXPECT_EQ(str, "bbcd");
+
+  str = "abcd";
+  replaceStringInPlace(str, "a", "");
+  EXPECT_EQ(str, "bcd");
```
Comment (mgautierfr) [inline test/tools-test.cpp:107]: We may add this test in `replaceStringInPlaceOnce` test also.

File: test/meson.build [hunk 8]
```diff
@@ -1,4 +1,5 @@
 gtest_dep = dependency('gtest', main:true, fallback:['gtest', 'gtest_main_dep'], required:false)
+icu_dep = dependency('icu-i18n', static:static_linkage)
```
[the remark this answers is not in the document]
Comment (asashnov) [reply test/meson.build:2]: I looked carefully through https://mesonbuild.com/Dependencies.html  and didn't found that a dependency can be associated with a source file. Dependencies always specified to the exe file only.

libicu is required by libzim, so maybe we should just link our tests executable with libzim, as we do it in tools directory:

```
executable('zimpatch', 'zimpatch.cpp',
  dependencies: libzim_dep,
  install: true)
```

[the remark this answers is not in the document]
Comment (asashnov) [reply test/meson.build:2]: Extra favicon.png is removed (from 'data/...', leaved in 'test/data/..')

File: test/tools-test.cpp [hunk 9]
```diff
@@ -0,0 +1,141 @@
+#include "gtest/gtest.h"
+
+#include "../src/tools.h"
+
+static std::string get_tests_data_dir()
+{
+  return "data";
+}
```
[the remark this answers is not in the document]
Comment (asashnov) [reply test/tools-test.cpp]: I wanted to leave it for future flexibility, but indeed, it's better to keep things as simple as it can be. Removed.

File: test/tools-test.cpp (hunk 5 above)
Comment (asashnov) [reply test/tools-test.cpp]: Probably, no, as this will require extra lines of code. And favicon.png is used anyway in further tests (#143)

File: test/meson.build (hunk 8 above)
[the remark this answers is not in the document]
Comment (mgautierfr) [reply test/meson.build:2]: icu is used in tools (https://github.com/openzim/zim-tools/blob/master/src/tools.cpp#L302) so it is not a requirement of libzim (in fact, it is but we cannot rely on this).
All binaries that use `removeAccents` need to link with icu. Not only the tests.
So we need to move the icu dependency global and make "real" binaries linking with it.

On the tests side, it is simpler to make all test linking with it (as it is done with other deps).

For now binaries are working because libzim also depend on it (if compiled with xapian). But we should not rely on this side effect.

[the remark this answers is not in the document]
Comment (asashnov) [reply test/meson.build]: Actually, we don't need libicu and 'removeAccents' function in zim-tools at all.
Seems it was unused copy/paste from libzim. I just removed `removeAccents()` code and test.
Now we don't need a special case for 'tools-test'.

[the remark this answers is not in the document]
Comment (asashnov) [reply test/meson.build]: `removeAccents()` seems to be unused copy/paste from libzim. So I just removed it.

File: src/tools.cpp [hunk 10]
```diff
@@ -294,20 +295,6 @@ std::string getNamespaceForMimeType(const std::string& mimeType, bool uniqueName
   }
 }
 
-
-std::string removeAccents(const std::string& text)
```
Comment (mgautierfr) [inline src/tools.cpp:298]: We must remove the declaration also.

**Comment by kelson42** [comment]
@MiguelRocha Good for you as well?

File: src/tools.cpp [hunk 11]
```diff
@@ -254,6 +252,9 @@ void replaceStringInPlace(std::string& subject,
                           const std::string& search,
                           const std::string& replace)
 {
+  if (search.empty())
```
Comment (MiguelRocha) [inline src/tools.cpp:255]: If the search is empty if means that we replace nothing. It should return the original string here.

File: test/tools-test.cpp [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+std::string directoryPath;
+magic_t magic;
+bool inflateHtmlFlag = false;
+bool isVerbose() { return false; }
+
+TEST(CommonTools, getFileSize)
+{
+  std::string fn = "data/minimal-content/favicon.png";
+
+  system("pwd");
```
Comment (MiguelRocha) [inline test/tools-test.cpp:15]: this system() syscall was probably for debugging?

File: test/tools-test.cpp [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+  EXPECT_EQ(res[0], '\0');
+
+  src = "%ff";
+  res = decodeUrl(src);
+  EXPECT_EQ(res.size(), 1);
+  EXPECT_EQ(res[0], '\xff');
+
+
+  // test corner cases
+  src = "";
```
Comment (MiguelRocha) [inline test/tools-test.cpp:51]: Here we could have a map of (original string, expected string) and then run it on a loop.

_pseudo code_
```
unordered_map<string, string> = expectationsMap = {  {"%", "%"}, {"%2", "%2"}, ..., ... };

for (auto p : expectationsMap) EXCPECT_EQ(p.first, p.second);

```

But its just a suggestion. No need to implement it for approval.

(I later noticed that there are other places in the code that this could be done)

File: test/tools-test.cpp (hunk 12 above)
Comment (asashnov) [reply test/tools-test.cpp:15]: Yes, it was for debug, removed. Excuse me.

File: test/tools-test.cpp (hunk 13 above)
Comment (asashnov) [reply test/tools-test.cpp]: Good idea to have at least one example in the code. One test was rewritten.

File: src/tools.cpp (hunk 11 above)
Comment (MiguelRocha) [reply src/tools.cpp:255]: True. Sorry for the confusion.

**Review by MiguelRocha — APPROVED** [review]
